keps/sig-apps/3017-pod-healthy-policy-for-pdb/README.md

```diff
@@ -306,9 +307,10 @@ We expect no non-infra related flakes in the last month as a GA graduation crite
 
 - Feature gate enabled by default.
 - Existing E2E and conformance tests passing.
-- Consider whether we want to keep the `spec.unhealthyPodEvictionPolicy` field null by default when not specified
-  or default it to `IfHealthyBudget` value. Both of these should preserve original behavior and handle null
-  values. This should be tested and documented.
+- As `IfHealthyBudget` is generally the less desirable option, it would be good to inform users that they are using it.
+  We can achieve this by defaulting `spec.unhealthyPodEvictionPolicy` to a `IfHealthyBudget` value. 
+  This might prompt users to re-evaluate the value and set it to `AlwaysAllow` which is preferable for most of the applications.
+  The default value should preserve original behavior and handle null values. This should be tested and documented.
```

Comment:
I'm not sure if defaulting will help with the adoption or rise awareness, @wojtek-t also pointed out that [here](https://github.com/kubernetes/enhancements/pull/3777/files#r1094534479) that it might cause issues for existing CI/CD flows where users have predefined PDBs with nil values for those fields which would cause them being pushed to the server. I'm inclined to leave it as is, for now. 

Do we have a precedence before for changing defaults in similar situations? If we decide to do that, I'd probably go the path of adding a warning for nil values and setting the default in the next release, to give users one release to adopt, maybe?

Reply:
there's plenty of precedence for defaulting of newly introduced fields, conditional on their feature being enabled, which is what is being considered here:

```
git grep 'Gate.Enabled' -- 'pkg/apis/*defaults.go'
pkg/apis/apps/v1/defaults.go:           if utilfeature.DefaultFeatureGate.Enabled(features.MaxUnavailableStatefulSet) {
pkg/apis/apps/v1/defaults.go:   if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
pkg/apis/apps/v1beta1/defaults.go:      if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
pkg/apis/apps/v1beta1/defaults.go:              if utilfeature.DefaultFeatureGate.Enabled(features.MaxUnavailableStatefulSet) {
pkg/apis/apps/v1beta2/defaults.go:              if utilfeature.DefaultFeatureGate.Enabled(features.MaxUnavailableStatefulSet) {
pkg/apis/apps/v1beta2/defaults.go:      if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
pkg/apis/storage/v1/defaults.go:        if obj.Spec.SELinuxMount == nil && utilfeature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
pkg/apis/storage/v1beta1/defaults.go:   if obj.Spec.SELinuxMount == nil && utilfeature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
```

that said, there's also evidence that some clients don't handle defaulted fields correctly / gracefully in general (including the kubelet choking on new defaulted fields inside podspec - https://github.com/kubernetes/kubernetes/issues/63814, or clients doing diffs being unhappy with defaulted fields - https://github.com/argoproj/argo-cd/issues/11143), so unless there's a strong reason to default, I'd probably leave the field `nil` by default and document which explicit option that corresponds to

Reply:
> so unless there's a strong reason to default, I'd probably leave the field nil by default and document which explicit option that corresponds to

+1 (unless there are other reasons not mentioned here).

Reply:
ack, updated the KEP to keep the field nil by default